core/query/query.ml

```diff
@@ -713,179 +810,112 @@ struct
+                                  | _ -> assert false
+                          in
+                            Q.For (None, gs, os @ os', body)
+                      | _ -> assert false
+                  end
         end
-  (* simple optimisations *)
-  and reduce_and (a, b) =
-    let open Q in
-    match a, b with
-      | Constant (Constant.Bool true), x
-      | x, Constant (Constant.Bool true)
-      | (Constant (Constant.Bool false) as x), _
-      | _, (Constant (Constant.Bool false) as x) -> x
-      | _ -> Apply ("&&", [a; b])
-  and reduce_or (a, b) =
-    let open Q in
-    match a, b with
-      | (Constant (Constant.Bool true) as x), _
-      | _, (Constant (Constant.Bool true) as x)
-      | Constant (Constant.Bool false), x
-      | x, Constant (Constant.Bool false) -> x
-      | _ -> Apply ("||", [a; b])
-  and reduce_not a =
-    let open Q in
-    match a with
-      | Constant (Constant.Bool false) -> Constant (Constant.Bool true)
-      | Constant (Constant.Bool true)  -> Constant (Constant.Bool false)
-      | _                       -> Apply ("not", [a])
-  and reduce_eq (a, b) =
-    let open Q in
-    let bool x = Constant (Constant.Bool x) in
-    let eq_constant =
-      function
-        | (Constant.Bool a  , Constant.Bool b)   -> bool (a = b)
-        | (Constant.Int a   , Constant.Int b)    -> bool (a = b)
-        | (Constant.Float a , Constant.Float b)  -> bool (a = b)
-        | (Constant.Char a  , Constant.Char b)   -> bool (a = b)
-        | (Constant.String a, Constant.String b) -> bool (a = b)
-        | (a, b)                 -> Apply ("==", [Constant a; Constant b])
-    in
-      match a, b with
-        | (Constant a, Constant b) -> eq_constant (a, b)
-        | (Variant (s1, a), Variant (s2, b)) ->
-          if s1 <> s2 then
-            Constant (Constant.Bool false)
-          else
-            reduce_eq (a, b)
-        | (Record lfields, Record rfields) ->
-          List.fold_right2
-            (fun (_, v1) (_, v2) e ->
-              reduce_and (reduce_eq (v1, v2), e))
-            (StringMap.to_alist lfields)
-            (StringMap.to_alist rfields)
-            (Constant (Constant.Bool true))
-        | (a, b) -> Apply ("==", [a; b])
+    | Q.Primitive "not", [v] ->
+      Q.reduce_not (v)
+    | Q.Primitive "&&", [v; w] ->
+      Q.reduce_and (v, w)
+    | Q.Primitive "||", [v; w] ->
+      Q.reduce_or (v, w)
+    | Q.Primitive "==", [v; w] ->
+      Q.reduce_eq (v, w)
+    | Q.Primitive "stringToXml", [v] ->
+      Q.Singleton (Q.XML (Value.Text (Q.unbox_string v)))
+    | Q.Primitive f, args ->
+        Q.Apply (Q.Primitive f, args)
+    | Q.If (c, t, e), args ->
+        Q.reduce_if_condition (c, apply env (t, args), apply env (e, args))
+    | Q.Apply (f, args), args' ->
+        apply env (f, args @ args')
+    | t, _ -> query_error "Application of non-function: %s" (string_of_t t)
+
+  and norm_comp env c = norm env (computation env c)
+
+(*
```

Comment:
Can this be cleaned up?

Reply:
I can inline it. Personally, though, I think the code is more readable this way. Does it even make any difference after compiling?

Reply:
"This" in my question was about the comment starting on line 903, not a suggestion to inline `norm_comp`.

Reply:
Ahhh I see. Ok.